### PR TITLE
RSDK-1155 Experimental WebRTC H264 RTP Passthrough 

### PR DESCRIPTION
**Pull request by nicksanford**
Demo uses:
https://github.com/erh/viamrtsp/pull/7

TL;DR:
I've tested that this doesn't affect existing cameras & modules (see the test cases below). I have not verified that the RTP passthrough code path is bug free (and there are known issues listed below).

I believe we can merge this as is as the new RTP passthrough code path will only be exercised if both the the module & camera model support it. AKA it is feature flagged at the module model level.

### Follow ups:
1. Adding RTP passthrough to fake camera
2. Adding module tests for the RTP passthrough code path
3. Manually testing `Configuration: passthrough viam-server <>  passthrough module`
4. Addressing the known limitations / future work

## Changes:

1. Add `camera.VideoCodecStreamSource` interface
5. Add `VideoCodecStreamSource(ctx context.Context) (VideoCodecStreamSource, error)` method to `camera.VideoSource` interface
6. Implement `VideoCodecStreamSource` method on camera client which returns `nil` if there is no peer connection & otherwise returns itself.
7. Implement `VideoCodecStreamSource` interface methods 
8. Implement stub `VideoCodecStreamSource` method for all other builtin cameras 
9. Update `camera.NewVideoSourceFromReader` to detect if the `reader gostream.VideoReader` parameter implements `camera.VideoCodecStreamSource` & if so return it from it from the `VideoCodecStreamSource`  method.
10. Update [gortsplib](https://github.com/bluenviron/gortsplib) from v2 to v4
11. Update https://github.com/pion/rtp from v1.8.2 -> v1.8.3
12. Make module.go detect all camera models in the module that return non nil values from their `VideoCodecStreamSource` method and accept the resource names of those cameras as streams.
13. Change module.go's `ListStream` method to return the names of the cameras detected on point 9
14. Change module.go's `AddStream` method to write RTP packets from the camera's `SubscribeRTP` method's callback to viam servert's WebRTC peer connection's track local static object.
15. Change  module.go's `RemoveStream` method  to call `Unsubscribe` on the camera & remove the track from the viam-server peer connection
16. Change `web_c.go` to detect if the camera behind a gostream video source supports h264 passthrough (must return non nil value from `VideoCodecStreamSource` method & call to `SubscribeRTP` must succeed). If supported, h264 passthrough will be used for callers of viam server's `AddStream` and `RemoveStream` methods. If unsupported, the current `GetImage` / gostream code path is used. If subsequent SubscribeRTP calls fail, will fall back to GetImage code path. Once GetImage code path begins, it will be used for the lifetime of the camera model name within a given viam-server instance (this is due to the design of the GoStream `StreamVideoSource` function.

### Note:
As currently implemented, this PR makes the camera client responsible for multiplexing `AddStream` and `RemoveStream` calls from different to viam-server peer connections (such as different browsers) for the same camera model instance on to a single `AddStream` call on a module or a remote.

This logic needs to live somewhere in viam-server (maybe there is a better place for it to live) because the module only gets a single peer connection with viam-server & as a result there is no way for the module to differentiate between `AddStream` or `RemoveStream` calls for the same camera model instance. That information is not expressible with the stream proto api we have today.

### Definition of terms:
1.  Ungraceful recovery: stream stops, no error message, user needs to toggle RC card UI element to restart stream)
2. Graceful recovery: stream stops & then starts again w/o user needing to take any action.

### Known Limitations / future work:
1. Graceful recovery is not working for h264 passthrough streams or streams that go from h264 passthrough -> GetImage path
2. Graceful recovery is not working for h264 passthrough streams after reconfiguration 
3. Once a camera model is using the GetImage code path (non-passthrough) it continues using non-passthrough until either its name changes or viam server reboots
4. Some of the passthrough code paths have false positive error logs (particualarly around unsubscribing from passthrough camera streams when resources close / reconfigure). 
5. I can't test whether or not this break module reconfiguration as the main branch of viam server currently has a broken module configuration code path: https://viam.atlassian.net/browse/RSDK-7173
6. Module process crashes leave passthrough cameras in zombie states
7. The camera RC card sometimes fails with an exception (including on viam-server stable `v0.24.0`) which requries a browser refresh to recover: https://viam.atlassian.net/browse/RSDK-7180 I believe this is an existing UI bug.

# Testing Checklist:

## Configuration: passthrough viam-server <> passthrough module:

Acceptance criteria for each test cases:
1. Open & close camera streams in app.viam.com RC UI & confirm video starts & stops
1. Reconfigure the camera model (change attrs) & confirm that it stops the video ( both browser & server side) and that closing & reopening the video feed results in receiving the post reconfigure video
1. Remove camera(s) under test while streaming video confirm only that expected camera(s) is/are removed from the RC card, and that the video(s) stop(s) ( both browser & server side)
1. Add a camera under test & confirm it configures properly & is suable (passes 1)
1. Remove a camera while not streaming video confirm only that camera is removed from the RC card
1. Rename a camera & confirm it behaves the same as being reconfigured
1. When you shut down viam server, confirm it terminates cleanly

## Configuration: passthrough viam-server <>  passthrough module:

This will be fully tested post merge.

- [x] Single camera (passthrough) single browser tab
- [x] Single camera (passthrough) single browser tab - revert model to non passthrough config
- [x] Single camera (non-passthrough) single browser tab
- [ ] Single camera (passthrough) multiple browser tabs: fails 2 & 3 until all webrtc client connections to viam-server either close or call `RemoveStream` before calling `AddStream`
- [ ] Single camera (passthrough) multiple browser tabs - revert model to non-passthrough config: BROKEN - need to refresh all browser sessions before it recovers
- [ ] Single camera (non-passthrough) multiple browser tabs
- [ ] Multiple cameras (passthrough) single browser tab  - revert model to non passthrough config
- [ ] Multiple cameras (mix passthrough & non-passthrough) single browser tab 
- [ ] Multiple cameras (non-passthrough) single browser tab 
- [ ] Multiple cameras (passthrough) multiple browser tabs
- [ ] Multiple cameras (mix passthrough & non-passthrough) multiple browser tabs
- [ ] Multiple cameras (non-passthrough) multiple browser tabs 
- [ ] Camera module under test crashing is able to recover
- [ ] Camera module under test crashing is able to recover gracefully 
- [ ] Unrelated module crashing is able to recover
- [ ] Unrelated module crashing is able to recover gracefully

## Configuration: passthrough viam-server <> non -passthrough module: (both module that supports passthrough with model that does not, a go module that does not support passthrough & a python module that does not support passthrough)
- [x] Single camera (non-passthrough) single browser tab
- [x] Single camera (non-passthrough) multiple browser tabs
- [x] Multiple cameras (non-passthrough) single browser tab 
- [x] Multiple cameras (non-passthrough) multiple browser tabs 
- [x] Camera module under test crashing is able to recover
- [x] Camera module under test crashing is able to recover gracefully 
- [x] Unrelated module crashing is able to recover
- [x] Unrelated module crashing is able to recover gracefully

## Configuration: non passthrough viam-server <> passthrough module `viam-server v0.24.0`
- [x] Single camera (non-passthrough) single browser tab
- [x] Single camera (non-passthrough) multiple browser tabs
- [x] Multiple cameras (non-passthrough) single browser tab 
- [x] Multiple cameras (non-passthrough) multiple browser tabs 
- [x] Camera module under test crashing is able to recover
- [x] Camera module under test crashing is able to recover gracefully 
- [x] Unrelated module crashing is able to recover
- [x] Unrelated module crashing is able to recover gracefully

### Test config:
```json
{
  "components": [
    {
      "attributes": {
        "rtsp_address": "rtsp://localhost:8554/live1",
        "h264_passthrough": false
      },
      "depends_on": [],
      "model": "erh:viamrtsp:rtsp-h264",
      "name": "viamrtsp-rtsp-h264-recording1",
      "namespace": "rdk",
      "type": "camera"
    },
    {
      "depends_on": [],
      "model": "erh:viamrtsp:rtsp-h264",
      "name": "viamrtsp-rtsp-h264-recording-no-passthrough1",
      "namespace": "rdk",
      "type": "camera",
      "attributes": {
        "rtsp_address": "rtsp://localhost:8554/live1"
      }
    },
    {
      "name": "dan-module1",
      "namespace": "rdk",
      "type": "camera",
      "attributes": {
        "animated": true
      },
      "depends_on": [],
      "model": "dang:camera:fake"
    },
    {
      "namespace": "rdk",
      "type": "camera",
      "attributes": {
        "camera_name": "builtin-fake1"
      },
      "depends_on": [
        "builtin-fake1"
      ],
      "model": "viam-labs:camera:overlay-fps",
      "name": "fps-builtin-fake1"
    },
    {
      "depends_on": [
        "viamrtsp-rtsp-h264-recording1"
      ],
      "model": "viam-labs:camera:overlay-fps",
      "name": "fps-h2641",
      "namespace": "rdk",
      "type": "camera",
      "attributes": {
        "camera_name": "viamrtsp-rtsp-h264-recording1"
      }
    },
    {
      "model": "fake",
      "name": "builtin-fake1",
      "namespace": "rdk",
      "type": "camera",
      "attributes": {
        "animated": true
      },
      "depends_on": []
    }
  ],
  "agent_config": {
    "subsystems": {
      "agent-provisioning": {
        "disable_subsystem": false,
        "release_channel": "stable",
        "pin_version": "",
        "pin_url": ""
      },
      "agent-syscfg": {
        "pin_url": "",
        "disable_subsystem": false,
        "release_channel": "stable",
        "pin_version": ""
      },
      "viam-agent": {
        "pin_version": "",
        "pin_url": "",
        "disable_subsystem": false,
        "release_channel": "stable"
      },
      "viam-server": {
        "pin_version": "",
        "pin_url": "",
        "disable_subsystem": false,
        "release_channel": "stable"
      }
    }
  },
  "modules": [
    {
      "name": "viamrtsp",
      "type": "local",
      "executable_path": "/Users/nicksanford/code/viamrtsp/bin/viamrtsp"
    },
    {
      "executable_path": "/Users/nicksanford/code/robotModules/modmain",
      "name": "cameraWebRTC",
      "type": "local"
    },
    {
      "module_id": "viam-labs:overlay-fps",
      "name": "viam-labs_overlay-fps1",
      "type": "registry",
      "version": "0.0.4"
    }
  ]
}
```

### CODE REVIEW

File: module/modmanager/manager.go [hunk 1]
```diff
@@ -1146,3 +1183,34 @@ func getModuleDataParentDirectory(options modmanageroptions.Options) string {
...
+}
+
+func connect(pc *webrtc.PeerConnection, encodedAnswer string) error {
+	answer := webrtc.SessionDescription{}
+	if err := modlib.DecodeSDP(encodedAnswer, &answer); err != nil {
+		return err
+	}
+
+	if err := pc.SetRemoteDescription(answer); err != nil {
+		return nil // nolint:nilerr
```
Comment (nicksanford) [inline module/modmanager/manager.go:1212]: I'm not sure why we don't return this error

File: testutils/file_utils.go [hunk 2]
```diff
@@ -21,7 +22,13 @@ func BuildTempModule(tb testing.TB, dir string) string {
 	builder := exec.Command("go", "build", "-o", modPath, ".")
 	builder.Dir = utils.ResolveFile(dir)
 	out, err := builder.CombinedOutput()
-	if len(out) != 0 {
+	// NOTE (Nick S): Workaround for the tickets below:
+	// https://viam.atlassian.net/browse/RSDK-7145
+	// https://viam.atlassian.net/browse/RSDK-7144
+	// Don't fail build if platform has known compile warnings (due to C deps)
+	isPlatformWithKnownCompileWarnings := runtime.GOARCH == "arm" || runtime.GOOS == "darwin"
```
Comment (nicksanford) [inline testutils/file_utils.go:29]: This is needed for the tests to pass on osx & 32 bit linux

File: components/camera/stream_subscription.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+// Start starts the subscription routine.
+func (w *StreamSubscription) Start() {
+	w.wg.Add(1)
+	utils.ManagedGo(w.run, w.wg.Done)
+	<-w.started
+}
+
+// Close closes the subscription routine.
+func (w *StreamSubscription) Close() {
+	defer w.wg.Wait()
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/stream_subscription.go]: agreed updated

File: components/camera/stream_subscription.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+// ID returns the id of the StreamSubscription.
+func (w *StreamSubscription) ID() uuid.UUID {
+	return w.id
+}
+
+// Start starts the subscription routine.
+func (w *StreamSubscription) Start() {
+	w.wg.Add(1)
+	utils.ManagedGo(w.run, w.wg.Done)
+	<-w.started
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/stream_subscription.go]: you are right, removed

File: robot/web/web_c.go [hunk 5]
```diff
@@ -259,16 +268,131 @@ func (svc *webService) propertiesFromStream(ctx context.Context, stream gostream
...
+			MaxSleep:  2 * time.Second,
+			Cooldown:  5 * time.Second,
+		}
+		err := webstream.StreamVideoSource(streamVideoCtx, source, stream, opts, svc.logger)
+		if utils.FilterOutError(err, context.Canceled) != nil {
+			svc.logger.CErrorw(streamVideoCtx, "error streaming", "error", err, "name", stream.Name())
+		}
+	}, svc.webWorkers.Done)
+
+	<-goroutineStartedCtx.Done()
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply robot/web/web_c.go]: you are right, removed

File: module/module.go [hunk 6]
```diff
@@ -156,6 +164,7 @@ type Module struct {
 	logger                  logging.Logger
 	mu                      sync.Mutex
 	activeResourceStreams   map[resource.Name]peerResourceState
+	streamSourceByName      map[resource.Name]camera.VideoCodecStreamSource
```
Comment (hexbabe) [inline module/module.go:167]: Q: why not embed `streamSource` in `peerResourceState`?

**Review by hexbabe — COMMENTED** [review]
Don't have too many comments here! Mainly a lot of reading and learning since I haven't worked so much on the RTSP/streaming side. Let me know if you need hardware to test other cameras/our 3D cameras.

File: module/module.go (hunk 6 above)
Comment (nicksanford) [reply module/module.go:167]: Because `streamSourceByName` contains all cameras in the module. `activeResourceStreams` only contains the cameras which are currently streaming video.

File: components/camera/replaypcd/replaypcd_test.go [hunk 7]
```diff
@@ -821,3 +821,21 @@ func TestReplayPCDReconfigure(t *testing.T) {
 
 	test.That(t, serverClose(), test.ShouldBeNil)
 }
+
+func TestVideoCodecStreamSource(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/replaypcd/replaypcd_test.go:825]: We are testing that the `VideoCodecStreamSource` returns unimplemented for this camera model. Not all `VideoSource` camera models will return unimplemented for this method but this one does.

File: components/camera/client.go [hunk 8]
```diff
@@ -318,13 +354,206 @@ func (c *client) DoCommand(ctx context.Context, cmd map[string]interface{}) (map
 // We will call `Close` on the camera client when we detect the disconnection to remove
 // active streams but then reuse the client when the connection is re-established.
 func (c *client) Close(ctx context.Context) error {
+	_, span := trace.StartSpan(ctx, "camera::client::Close")
+	defer span.End()
+
 	c.mu.Lock()
+	// NOTE: we are intenionally releaseing the lock before we wait for
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/client.go]: done

File: components/camera/camera_test.go [hunk 9]
```diff
@@ -87,6 +110,27 @@ func TestNewPinholeModelWithBrownConradyDistortion(t *testing.T) {
 	test.That(t, pinholeCameraModel4.Distortion, test.ShouldBeNil)
 }
 
+func TestNewVideoSourceFromReader(t *testing.T) {
+	t.Run("if reader is a VideoCodecStreamSource, returns a VideoSource "+
+		"who's VideoCodecStreamSource method returns the reader", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/camera_test.go]: done

File: components/camera/client.go [hunk 10]
```diff
@@ -27,17 +36,41 @@ import (
 	"go.viam.com/rdk/utils"
 )
 
+var (
+	// ErrNoPeerConnection indicates there was no peer connection.
+	ErrNoPeerConnection = errors.New("ErrNoPeerConnection")
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/client.go]: done

File: components/camera/stream_subscription.go [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+var (
+	// ErrQueueFull indicates the StreamSubscription's queue is full and that
+	// the callback passed to Publish will not be executed.
+	ErrQueueFull = errors.New("StreamSubscription Publish queue full")
+	// ErrClosed indicates the StreamSubscription's is not running.
+	ErrClosed = errors.New("StreamSubscription has been closed")
+	// ErrNegativeQueueSize indicates that the StreamSubscription queueSize
+	// can't be less than 0.
+	ErrNegativeQueueSize = errors.New("ErrNegativeQueueSize")
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/stream_subscription.go]: done

**Review by dgottlieb — APPROVED** [review]

File: components/camera/client.go [hunk 12]
```diff
@@ -5,19 +5,28 @@ import (
...
+	"github.com/pion/webrtc/v3"
 	"github.com/pkg/errors"
 	"go.opencensus.io/trace"
+	"go.uber.org/multierr"
 	pb "go.viam.com/api/component/camera/v1"
+	streampb "go.viam.com/api/stream/v1"
 	goutils "go.viam.com/utils"
 	goprotoutils "go.viam.com/utils/protoutils"
 	"go.viam.com/utils/rpc"
+	"golang.org/x/exp/slices"
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/client.go:23]: @dgottlieb  FYI I need to add back exp/slices

The linter in CI appers to be non-deterministic when it comes to where `"slices"` should be in the import list & I see other RDK packages use `exp/slices`

https://github.com/viamrobotics/rdk/actions/runs/8544406639/job/23410383689?pr=3723
https://github.com/viamrobotics/rdk/actions/runs/8544116872/job/23409464242#step:7:89

[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/client.go:23]: Bug filed to fix that issue: https://viam.atlassian.net/browse/RSDK-7202

[the remark this answers is not in the document]
Comment (dgottlieb) [reply components/camera/client.go:23]: I'm not sure what the linter is doing wrong? Are you sure you ran it/goimports locally? It seems from the patch ran, the new "slices" import was still in the "old" spot (with all of the non-stdlib imports). But I would expect the patch to have moved it to the stdlib side of the imports.

image

[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/client.go:23]: The linter (run in cannon) was swapping the location of `"slices"`  every time it was run: 
Between every commit I ran `make lint-go` in canon & the linter swaps the location of `"slices"`
https://github.com/viamrobotics/rdk/pull/3766/commits

[the remark this answers is not in the document]
Comment (dgottlieb) [reply components/camera/client.go:23]: Oh, I see now. The term non-deterministic was confusing me. Strictly speaking, the inputs are* different. This is an idempotency violation where `f(x) != f(f(x))`. 

The diff from the error I clicked through moving it into the "right" spot seemed correct. But now I see when run on the input where slices is in the right spot seems to be moving it to the wrong spot.

**Comment by nicksanford** [comment]
@dgottlieb  @randhid I did more manual testing and found that if there is a go module on `v0.24.1` or `v0.24.0` of the RDK (stable & stable-1) with camera models, those camera models won't be able to stream video to the RC UI after this change. This is b/c after this change, viam-server will falsely believe. that they support RTP passthrough when they in fact do not. 

This is b/c the current [module.go](https://github.com/viamrobotics/rdk/blob/1caac5375e5b028db60696bbd88eabe867040a96/module/module.go#L674) will return success from AddStream & then never write to the track which leaves viam-server always waiting for RTP packets. This is a mistake I made in a previous update to module.go.

The options I see are:
1. We live with this risk. I recommend this path especially if we don't know of any go camera modules other than https://github.com/erh/viamrtsp. I've confirmed the erh/viamrtsp module (both main branch & the currently deployed version on the registry) it is not affected by this problem. We would just need to make sure we don't update erh/viamrtsp to use the RDK versions with the problem.
1. I could look into changing the camera client SubscribeRTP method to wait a finite amount of time for the OnTrack callback to fire. That would prevent the waiting forever scenario but would likely take more time than just today to fix.

Are we aware of any go camera modules other than https://github.com/erh/viamrtsp that might be affected by this?



I would like to recommend we still move forward with this change

**Comment by dgottlieb** [comment]
I'd also recommend moving forward with the change.

**Review by randhid — APPROVED** [review]
Mostly extremely optional comments trying to track the control logic.

The comments are trying to point out what's important in establishing a PeerConenction with a Module implementing a VideoCameraStreamSoruce, making the resource graph aware that is is a peer connection that allows rtp packet passthrough, bookkeeping so that multiple clients requesting videostreams from viam server form that peer connection do not establish more peer connection with that module, and passing those packets to the clients that want Vide, such as the rc card or a direct video stream to that camera. I probably got some things wrong. 

I agree we should merge this in!

File: components/camera/client.go [hunk 13]
```diff
@@ -318,13 +352,211 @@ func (c *client) DoCommand(ctx context.Context, cmd map[string]interface{}) (map
...
+			"name", c.Name(), "subAndCallbackByID", c.subAndCallbackByID.String())
+	}()
+
+	c.subAndCallbackByID[sub.ID()] = subAndCallback{
+		cb:  func(p *rtp.Packet) error { return packetsCB([]*rtp.Packet{p}) },
+		sub: sub,
+	}
+
+	sub.Start()
+	sc.AddOnTrackSub(c.Name(), c.addOnTrackSubFunc)
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/client.go]: updated

File: components/camera/client.go [hunk 14]
```diff
@@ -318,13 +352,211 @@ func (c *client) DoCommand(ctx context.Context, cmd map[string]interface{}) (map
...
+	sub, err := NewStreamSubscription(bufferSize, onError)
+	if err != nil {
+		return uuid.Nil, err
+	}
+
+	if c.conn.PeerConn() == nil {
+		return uuid.Nil, ErrNoPeerConnection
+	}
+
+	sc, ok := c.conn.(*rdkgrpc.SharedConn)
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/client.go:407]: updated

File: module/module.go [hunk 15]
```diff
@@ -660,22 +736,49 @@ func (m *Module) AddStream(ctx context.Context, req *streampb.AddStreamRequest)
...
+		return &streampb.AddStreamResponse{}, nil
+	}
+
 	tlsRTP, err := webrtc.NewTrackLocalStaticRTP(webrtc.RTPCodecCapability{MimeType: "video/H264"}, "video", name.String())
 	if err != nil {
 		return nil, errors.Wrap(err, "error creating a new TrackLocalStaticRTP")
 	}
 
-	// TODO call remove track on error
+	subID, err := vcss.SubscribeRTP(ctx, rtpBufferSize, func(pkts []*rtp.Packet) error {
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply module/module.go:758]: updated

File: module/module.go [hunk 16]
```diff
@@ -438,8 +457,18 @@ func (m *Module) AddResource(ctx context.Context, req *pb.AddResourceRequest) (*
...
+	var vcss camera.VideoCodecStreamSource
+	if cam, ok := res.(camera.VideoSource); ok {
+		if v, err := cam.VideoCodecStreamSource(ctx); err == nil {
+			vcss = v
+		}
+	}
+
 	m.mu.Lock()
 	defer m.mu.Unlock()
+	if vcss != nil {
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply module/module.go:470]: updated

File: components/camera/client.go [hunk 17]
```diff
@@ -318,13 +352,211 @@ func (c *client) DoCommand(ctx context.Context, cmd map[string]interface{}) (map
...
+
+	c.subAndCallbackByID[sub.ID()] = subAndCallback{
+		cb:  func(p *rtp.Packet) error { return packetsCB([]*rtp.Packet{p}) },
+		sub: sub,
+	}
+
+	sub.Start()
+	sc.AddOnTrackSub(c.Name(), c.addOnTrackSubFunc)
+
+	if len(c.subAndCallbackByID) == 1 {
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/client.go:441]: updated

File: components/camera/replaypcd/replaypcd.go [hunk 18]
```diff
@@ -166,6 +166,10 @@ func newPCDCamera(
 	return cam, nil
 }
 
+func (replay *pcdCamera) VideoCodecStreamSource(ctx context.Context) (camera.VideoCodecStreamSource, error) {
+	return nil, errors.New("pcdCamera does not implement VideoCodecStreamSource")
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/replaypcd/replaypcd.go:170]: yes

File: components/camera/stream_subscription.go [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+	ErrNegativeQueueSize = errors.New("StreamSubscription queue size can't be negative")
+)
+
+// StreamSubscription executes the callbacks sent to Publish
+// in a single goroutine & drops Publish callbacks if the
+// buffer is full.
+// This is desirable behavior for streaming protocols where
+// dropping stale packets is desirable to minimize latency.
+type StreamSubscription struct {
+	buffer  *ringbuffer.RingBuffer
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/stream_subscription.go:49]: lol, I think that this can be something I take responsibility for figuring out. It is not obvious to me that it is actually a problem that we are using a not-a-ringbuffer here. I'd like to measure the problem before I callout something that might not be a problem

File: module/module.go [hunk 20]
```diff
@@ -510,17 +539,38 @@ func (m *Module) ReconfigureResource(ctx context.Context, req *pb.ReconfigureRes
 		m.logger.Error(err)
 	}
 
+	prs, ok := m.activeResourceStreams[res.Name()]
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply module/module.go:543]: done

File: components/camera/client.go [hunk 21]
```diff
@@ -318,13 +352,211 @@ func (c *client) DoCommand(ctx context.Context, cmd map[string]interface{}) (map
...
+	if !ok {
+		return uuid.Nil, ErrNoSharedPeerConnection
+	}
+	c.logger.CDebugw(ctx, "SubscribeRTP", "subID", sub.ID().String(), "name", c.Name(), "subAndCallbackByID", c.subAndCallbackByID.String())
+	defer func() {
+		c.logger.CDebugw(ctx, "SubscribeRTP after", "subID", sub.ID().String(),
+			"name", c.Name(), "subAndCallbackByID", c.subAndCallbackByID.String())
+	}()
+
+	c.subAndCallbackByID[sub.ID()] = subAndCallback{
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/client.go:420]: done

File: components/camera/client.go [hunk 22]
```diff
@@ -318,13 +352,211 @@ func (c *client) DoCommand(ctx context.Context, cmd map[string]interface{}) (map
...
+	if err := c.unsubscribeAll(); err != nil {
+		c.logger.CErrorw(ctx, "Close > unsubscribeAll", "err", err, "name", c.Name())
+		return err
+	}
+
+	return nil
+}
+
+func (c *client) VideoCodecStreamSource(ctx context.Context) (VideoCodecStreamSource, error) {
+	_, ok := c.conn.(*rdkgrpc.SharedConn)
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/client.go:382]: done

File: components/camera/client.go [hunk 23]
```diff
@@ -318,13 +352,211 @@ func (c *client) DoCommand(ctx context.Context, cmd map[string]interface{}) (map
...
+	defer c.activeBackgroundWorkers.Wait()
 	defer c.mu.Unlock()
 
 	if c.healthyClientCh != nil {
 		close(c.healthyClientCh)
 	}
-	c.activeBackgroundWorkers.Wait()
 	c.healthyClientCh = nil
+
+	if err := c.unsubscribeAll(); err != nil {
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply components/camera/client.go:371]: done

**Comment by hexbabe** [comment]
> Single camera (passthrough) single browser tab
 Single camera (passthrough) single browser tab - revert model to non passthrough config
 Single camera (non-passthrough) single browser tab
 Single camera (passthrough) multiple browser tabs: fails 2 & 3 until all webrtc client connections to viam-server either close or call RemoveStream before calling AddStream
 Single camera (passthrough) multiple browser tabs - revert model to non-passthrough config: BROKEN - need to refresh all browser sessions before it recovers
 Single camera (non-passthrough) multiple browser tabs
 Multiple cameras (passthrough) single browser tab - revert model to non passthrough config
 Multiple cameras (mix passthrough & non-passthrough) single browser tab
 Multiple cameras (non-passthrough) single browser tab
 Multiple cameras (passthrough) multiple browser tabs
 Multiple cameras (mix passthrough & non-passthrough) multiple browser tabs
 Multiple cameras (non-passthrough) multiple browser tabs
 Camera module under test crashing is able to recover
 Camera module under test crashing is able to recover gracefully
 Unrelated module crashing is able to recover
 Unrelated module crashing is able to recover gracefully
 
 is this testing list updated?

**Comment by nicksanford** [comment]
> > Single camera (passthrough) single browser tab
> > Single camera (passthrough) single browser tab - revert model to non passthrough config
> > Single camera (non-passthrough) single browser tab
> > Single camera (passthrough) multiple browser tabs: fails 2 & 3 until all webrtc client connections to viam-server either close or call RemoveStream before calling AddStream
> > Single camera (passthrough) multiple browser tabs - revert model to non-passthrough config: BROKEN - need to refresh all browser sessions before it recovers
> > Single camera (non-passthrough) multiple browser tabs
> > Multiple cameras (passthrough) single browser tab - revert model to non passthrough config
> > Multiple cameras (mix passthrough & non-passthrough) single browser tab
> > Multiple cameras (non-passthrough) single browser tab
> > Multiple cameras (passthrough) multiple browser tabs
> > Multiple cameras (mix passthrough & non-passthrough) multiple browser tabs
> > Multiple cameras (non-passthrough) multiple browser tabs
> > Camera module under test crashing is able to recover
> > Camera module under test crashing is able to recover gracefully
> > Unrelated module crashing is able to recover
> > Unrelated module crashing is able to recover gracefully
> 
> is this testing list up to date?

It is what I did prior to merging the PR
